userspace/libsinsp/parsers.cpp

```diff
@@ -5232,13 +5221,41 @@ void sinsp_parser::parse_container_json_evt(sinsp_evt *evt)
 			}
 		}
 
+		evt->set_tinfo_ref(container_info->get_tinfo(m_inspector));
+		evt->set_tinfo(evt->get_tinfo_ref().get());
+		container_info->set_lookup_status(static_cast<sinsp_container_lookup::state>(lookup_state.asUInt()));
+		switch(container_info->get_lookup_status())
+		{
+		// Preserve case where lookup_state == STARTED from scap file was set to FAILED
+		// or otherwise set to false.
+		case sinsp_container_lookup::state::FAILED:
+			break;
+		default:
+			{
+				// note: The container image is the most crucial field from a security incident response perspective. 
+				// We aim to raise the bar for successful container lookups. Conversely, pod sandboxes do not include  
+				// a container image in the API response.
+				if (!container_info->m_image.empty() || container_info->is_pod_sandbox())
+				{
+					// Marked as SUCCESSFUL if and only if the container image field present or 
+					// if it's a pod sandbox container
+					container_info->set_lookup_status(sinsp_container_lookup::state::SUCCESSFUL);
+				} else
+				{
+					container_info->set_lookup_status(sinsp_container_lookup::state::FAILED);
+				}
+			}
+		}
+
+		// note: Important, only set event filtering flag after all final state updates are done
 		if(!container_info->is_successful())
 		{
 			SINSP_DEBUG("Filtering container event for failed lookup of %s (but calling callbacks anyway)", container_info->m_id.c_str());
 			evt->set_filtered_out(true);
 		}
-		evt->set_tinfo_ref(container_info->get_tinfo(m_inspector));
-		evt->set_tinfo(evt->get_tinfo_ref().get());
+
+		// `add_container` also triggers the callbacks, therefore always continue even if the
+		// lookup_status was set to FAILED
```

Comment:
Another comment: This is not good in general. In another PR we should call a new method to trigger the callbacks when the status was set to FAILED and not add a failed container lookup to the cache ...  CC @therealbobo perhaps you wanna take that on? (As I need to reduce my PRs until June as I will be out a lot, after that can help again more)

Reply:
We want the failed lookups in the cache. Otherwise we'll retry them every time (on every syscall from a process in an affected container, whenever a lookup is not already in flight).

Retrying the lookups might not be a bad thing but it would need to happen in a much more controlled manner.

Reply:
Understood. Yes, if I may politely suggest, the current setup seems less than ideal, and I would say we should do something about it in the near future. See also https://github.com/falcosecurity/libs/issues/291

Reply:
I entirely agree, the current code is far from perfect. I know, I wrote it ;)

I'm all for updating container info over time if possible, it's just that dropping failed lookups is not the right approach. Let's split SUCCESSFUL into SUCCESSFUL_COMPLETE and SUCCESSFUL_INCOMPLETE and drive retries based on this (with some delays, backoff etc.)

Reply:
> I entirely agree, the current code is far from perfect. I know, I wrote it ;)
> 
> I'm all for updating container info over time if possible, it's just that dropping failed lookups is not the right approach. Let's split SUCCESSFUL into SUCCESSFUL_COMPLETE and SUCCESSFUL_INCOMPLETE and drive retries based on this (with some delays, backoff etc.)

Yes this seems like a great path forward. Created a ticket https://github.com/falcosecurity/libs/issues/1708.

Reply:
+1